kotsu/run.py

```diff
@@ -0,0 +1,60 @@
+"""Interface for running a registry of models on a registry of validations."""
+from typing import Tuple
+
+import logging
+import time
+
+from kotsu import store
+from kotsu.registration import ModelRegistry, ValidationRegistry
+
+
+logger = logging.getLogger(__name__)
+
+
+def run(
+    model_registry: ModelRegistry,
+    validation_registry: ValidationRegistry,
+    store_directory: str,
+    run_params: dict = {},
+):
+    """Run a registry of models on a registry of validations.
+
+    Args:
+        model_registry: A ModelRegistry containing the registry of models to be run through
+            validations.
+        validation_registry: A ValidationRegistry containing the registry of validations to run
+            each model through.
+        store_directory: A file path or URI location to store the validation results and any extra
+            output artefacts of the validations and models.
+        run_params: A dictionary of optional run parameters.
+    """
+    results_list = []
+    for validation_spec in validation_registry.all():
+        for model_spec in model_registry.all():
+            logger.info(f"Running validation - model: {validation_spec.id} - {model_spec.id}")
+            validation = validation_spec.make()
```

Comment:
Is there an argument for putting the validation_spec.make() outside of the loop over model specs? I would have thought that a single validation instance could be shared by all models. Also building the validation environment could be an expensive step so we'd want to avoid repeating unnecessarily (e.g. in your end_to_end test there is a load_dataset call within the validation callable which is the kind of thing that in certain cases could be time consuming. If this was put into the validation factory and validation_spec.make() was moved out of the loop over models, this would only need to be called once). If we were getting into distributing things it might need to look different but if doing everything serially like this I think reusing would make sense?

Reply:
Yeap this was something I was dwelling upon while writing. 

The two main competing thoughts were:
- Having a freshly instantiated `validation` for each model makes it easy to reason about the exact state of that instance and what it will do when given a model. It makes it more "dumb", in the sense that a `validation` doesn't have to handle multiple models, or clean up, or refresh etc. It just gets made, takes a model, sends its results, and then dies, and thereby avoids any possible insidious bugs where the state of the `validation` becomes "corrupted" after running on a model and its next results on models aren't correct. The thought here is to make it easier and safer for the user.*
- Using a single `validation` instance for all the models would be more efficient in some cases. 

*Given that users already have to be aware of stochasticity (specifically removing it), maybe it's not much more of a jump to required them to be aware of the state of the validation not being affected by running on a model.

Reply:
Also thought added to #19